modules/tinymce/src/themes/silver/test/ts/webdriver/components/urlinput/UrlInputTest.ts

```diff
@@ -0,0 +1,78 @@
+import { RealKeys, UiControls } from '@ephox/agar';
+import { GuiFactory, Representing, TestHelpers } from '@ephox/alloy';
+import { beforeEach, context, describe, it } from '@ephox/bedrock-client';
+import { Future, Optional } from '@ephox/katamari';
+import { SelectorFind, SugarDocument, Value } from '@ephox/sugar';
+import { assert } from 'chai';
+
+import { ApiUrlData } from 'tinymce/themes/silver/backstage/UrlInputBackstage';
+import { renderUrlInput } from 'tinymce/themes/silver/ui/dialog/UrlInput';
+
+import * as TestExtras from '../../../module/TestExtras';
+
+describe('webdriver.tinymce.themes.silver.components.urlinput.UrlInputTest', () => {
+  const helpers = TestExtras.bddSetup();
+
+  const hook = TestHelpers.GuiSetup.bddSetup((store, _doc, _body) => GuiFactory.build(
+    renderUrlInput({
+      label: Optional.some('UrlInput label'),
+      name: 'col1',
+      filetype: 'file',
+      enabled: true
+    }, helpers.backstage(), {
+      getHistory: (_fileType) => [],
+      addToHistory: (_url, _filetype) => store.adder('addToHistory')(),
+      getLinkInformation: () => Optional.none(),
+      getValidationHandler: () => Optional.none(),
+      getUrlPicker: (_filetype) => Optional.some((entry: ApiUrlData) => {
+        store.adder('urlpicker')();
+        return Future.pure({ value: 'http://tiny.cloud', meta: { before: entry.value }, fieldname: 'test' });
+      })
+    }, Optional.none())
+  ));
+
+  TestHelpers.GuiSetup.bddAddStyles(SugarDocument.getDocument(), [
+    '.tox-menu { background: white; }',
+    '.tox-collection__item--active { background: #cadbee }'
+  ]);
+
+  const getInput = () => {
+    const component = hook.component();
+    return component.getSystem().getByDom(
+      SelectorFind.descendant(component.element, 'input').getOrDie('Could not find input')
+    ).getOrDie();
+  };
+
+  const assertTrimSpaces = (inputKeys: any[], expectedOutput: string) =>
+    async () => {
+      const input = getInput();
+
+      await RealKeys.pSendKeysOn('body input', inputKeys);
```

Comment:
is `RealKeys` really the only way to trigger a form input event? That doesn't seem right. The more we rely on real keys the less tests we can run on our dev machines.

Reply:
I had told Hannes to do this from our chat last week... we cannot use any simulated APIs (such as those in Agar) here because the issue is that the browser itself trims all whitespace when using the `value` API on the input element. So if we used them then the input element would never end up with leading/trailing spaces anyways.